README.md
[![Build Status][travis-image]][travis-url]
# Serial Vault

A Go web service that digitally signs device assertion details.

The application can be run in three modes: signing service, admin service and system-user assertion services. All the web services
operate under unencrypted HTTP connections, so these should not be exposed to a public network
as-is. The services should be protected by web server front-end services, such as Apache, that
provide secure HTTPS connections. Also, the admin service does not include authentication nor 
authorisation, so this service will typically be made available on a restricted network with some 
authentication front-end on the web server e.g. SSO. Typically, the services will only be available 
on a restricted network at a factory, though, with additional security measures, the signing service 
could be made available on a public network.

Some deployment recommendations are [provided](docs/installation.md)

The service mode (signing or admin) is defined in the settings.yaml file. The
selected service should be accessible on port :8080 or :8081:
 - Signing/API Service: http://localhost:8080/v1/version
 - Admin/UI Service: http://localhost:8081/

The Admin service's CSRF protection sends a cookie over a secure channel. If the cookie is to be sent
over an insecure channel, it is needed to workaround it by setting the environment variable:
```bash
$ export CSRF_SECURE=disable
```
When modified that environment variable is set, consider that the current web session must be invalidated
in order to changes take effect. That could require a browser restart.
NEVER set this configuration in production environments.

## Install from Source
If you have a Go development environment set up, Go get it, we recommend at least Go v1.13 or higher.

  ```bash
  $ go get github.com/CanonicalLtd/serial-vault/...
  ```

### Configure it:
- Install PostgreSQL and create a database.
- Set up the config file, using ```settings.yaml``` as a guide.
- Create the database tables:
  ```bash
  $ cd serial-vault
  $ go get ./...
  $ go run cmd/serial-vault-admin/main.go database --config=/path/to/settings.yaml
  ```

Sample Serial Vault Configuration:
```
title: "Serial Vault"
logo: "/static/images/logo-ubuntu-white.svg"

# Path to the assets (${docRoot}/static)
docRoot: "."

# Backend database details
driver: "postgres"
datasource: "postgres://vault:vault@localhost:5432/vault?sslmode=disable"

keystore: "database"
keystoreSecret: "KEYSTORE_SECRET"

# Valid API keys
apiKeys:
  - API_KEY

# 32 bytes long key to protect server from cross site request forgery attacks
csrfAuthKey: "32_BYTES_LONG_CSRF_AUTH_KEY"
```

### Run it:
  ```bash
  $ cd $GOPATH/src/github.com/CanonicalLtd/serial-vault
  $ go run cmd/serial-vault/main.go -config=/path/to/settings.yaml -mode=signing
  ```

The application has an admin/UI service that can be run by using mode=admin.

## Deploy it with Juju
Juju greatly simplifies the deployment of the Serial Vault. A charm bundle is available
at the [charm store](https://jujucharms.com/u/canonical-solutions/serial-vault-bundle/), which deploys
everything apart from the Apache front-end units. There is an example of using Juju in the
[Deployment Guidelines](docs/installation.md).

## Try with docker

  ```bash
  $ git clone https://github.com/CanonicalLtd/serial-vault
  $ cd serial-vault/docker-compose
  $ docker-compose up
  # remove containers after try
  $ docker-compose kill && docker-compose rm
  ```

## Development Environment

### Contributing

The general workflow is [forking](https://help.github.com/en/github/getting-started-with-github/fork-a-repo) the Serial Vault GitHub repository, 
make changes in a branch and then create a [pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork).

#### Adding new golang dependency

`Serial Vault` uses `go mod` to manage its dependencies.

- Run `go get foo/bar` in the source folder to add the dependency to go.mod file.
- Run `go build ./...` to check that everything works.

To remove a dependency

- Edit your code and remove the import reference.
- Run `go mod tidy` in the source folder to remove dependency from go.mod file.

### Install Go
Follow the instructions to [install Go](https://golang.org/doc/install).

### Install the React development environment
#### Pre-requisites
- Install the build packages
```bash
sudo apt-get install build-essential libssl-dev
# For TPM2.0 (optional)
sudo apt-get install tpm2-tools
```

- Install NVM
Install the [Node Version Manager](https://github.com/creationix/nvm) that will allow a specific
version of Node.js to be installed. Follow the installation instructions.

- Install the latest stable Node.js and npm
The latest stable (LTS) version of Node can be found on the [Node website](nodejs.org).
```bash
# Overview of available commands
nvm help

# Install the latest stable version
nvm install lts/*

# Select the version to use
nvm ls
nvm use lts/*
```

- Install the nodejs dependencies
```bash
cd serial-vault
npm install
```

### Working with React

The frontend code can be found in `webapp-admin` directory.

#### Building static assets locally

```bash
# Select the version to use
cd webapp-admin/
nvm ls
nvm use lts/*
npm run build
```

#### Production static assets build process

Production build for the frontend part (javascript and css) is semi-automated and done with [GitHub Actions](https://github.com/features/actions). 
You can find the configuration for this process in `.github/workflows/nodejs.yml`. The build process starts automatically after the PR is approved 
and pushed to `master`. 

You can see the build process in [actions](https://github.com/CanonicalLtd/serial-vault/actions) tab of this project.
After the successful build the automation bot will create a PR with the generated build artifact (minified javascript code) in the `static/` 
directory of this project. These PRs can be merged manually.

#### Run the tests
```bash
npm test
```


## API Methods

### /v1/version (GET)
> Return the version of the serial vault service.

#### Output message
```json
{
  "version":"2.1-0",
}
```
- version: the version of the serial vault service (string)


### /v1/request-id (POST)
> Returns a nonce that is needed for the 'serial' request.

#### Output message
```json
{
  "request-id": "abc123456",
  "success": true,
  "message": ""
}
```
- success: whether the request was successful (bool)
- message: error message from the request (string)
- request-id: unique string that is needed for serial requests (string)

The request-id is a nonce that can only be used once and must be used before it expires (typically 600 seconds).

### /v1/serial (POST)
> Generate a serial assertion signed by the brand key.

Takes the details from the device as a serial-request assertion and generates a signed serial assertion.

#### Input message
The message must be the serial-request assertion format and is best generated using the snapd libraries.
```
type: serial-request
brand-id: System
model: Router 3400
device-key:
    WkUDQbqFCKZBPvKbwR...
request-id: abc123456
body-length: 10
sign-key-sha3-384: UytTqTvREVhx...

HW-DETAILS
serial: A1228ML

AcLBUgQAAQoABgUCV7R2C...
```
- brand-id: the Account ID of the manufacturer (string)
- model: the name of the device (string)
- device-key: the encoded type and public key of the device (string)
- request-id: the nonce returned from the /v1/nonce method (string)
- signature: the signed data
- serial: serial number of the device (string)

The HW-DETAILS are optional hardware details in YAML format, but must include the 'serial' tag as that is a mandatory
part of the serial assertion.

#### Output message
The method returns a signed serial assertion using the key from the vault.

### /v1/pivot (POST)
> Find the model pivot details for a device.

Takes the serial assertion from a device and returns the model pivot details for the device.

#### Input message
The serial assertion of the manufacturer's device.

#### Output message
The method returns details of the model pivot, to convert the device to a reseller model.

### /v1/pivotmodel (POST)
> Generate the model assertion for the pivoted model for a device.

Takes the serial assertion from a device and returns the model assertion for the pivoted model of the device.

#### Input message
The serial assertion of the manufacturer's device.

#### Output message
The method returns details of the model assertion of the pivoted model, to convert the device to a reseller model.

### /v1/pivotserial (POST)
> Generate the serial assertion for the pivoted model for a device.

Takes the serial assertion from a device and returns the serial assertion for the pivoted model of the device.

#### Input message
The serial assertion of the manufacturer's device.

#### Output message
The method returns details of the serial assertion of the pivoted model, to convert the device to a reseller model.

[travis-image]: https://travis-ci.org/CanonicalLtd/serial-vault.svg?branch=master
[travis-url]: https://travis-ci.org/CanonicalLtd/serial-vault
[actions]: https://github.com/CanonicalLtd/serial-vault/actions


ChangeLog
2020-03-23 serial-vault 2.5-0

  * Moving from govendor which is deprecated now to go mod for the dependency management.

  * Snapd version is updated to 2.44, all older versions were not compatible with Go mod.

  * All other external dependencies will keep the same pinned version.
    Note on versions: after the automated migration, all the pinned versions are moved from 
    `vendor/vendor.json` to `go.mod` file. If the external module does not use semantic versioning (SemVer) 
    in a format `v1.2.3` Go modules tooling will use `revision` and `revisionTime` from the pinned commit. 
    So the same git commit is pinned in`vendor/vendor.json` and in `go.mod` after the migration.
  
  * Updated documentation about how to use Go modules

  * Updated travis configuration to use Go modules

2017-09-05 serial-vault 2.0-0

  [ unknown ]
  * Added TravisCI support
    
  * See more information in merge proposal
    
  * See more information in merge proposal
    
  * Updated the README for the models API method and configuring the app.
    
  * Fix empty data checks for SignHandler
    
  * Added the basic web application to load the HTML template.
    
  * See more information in merge proposal
    
  * Validate the model against the accepted models before signing.
    
  * Add create, update and get methods for a model.
    
  * Resolve conflicts from merging master.
    
  * Use react-intl for i18n
    
  * API and UI to view, create and remove authorized public keys.
    
  * Resolve conflicts from merging master.
    
  * i18n for the user interface and API
    
  * Use Ubuntu-core assertions module for core operations
    
  * Keypair frontend
    
  * See more information in merge proposal
    
  * Show models that are linked to inactive keypairs.
    
  * See more information in merge proposal
    
  * Update dependencies and fixed code after Serial assertion renamed 
    
  * See more information in merge proposal
    
  * Use TPM 2.0 for keypair storage
    
  * See more information in merge proposal
    
  * Service refactor to split into a signing service and admin service
    
  * See more information in merge proposal
    
  * Add document root config setting to aid deployment.
    
  * See more information in merge proposal
    
  * Docker compose 1.5.2
    
  * Improve launch options when running as a snap
    
  * Check that the serial number and the device-key fingerprint are not re-used when signing assertions.
    
  * Update version to 0.4
    
  * Signing API protected by an API key
    
  * Set the version from a constant not the config file.
    
  * Update the snap version to 0.6
    
  * Formatted log messages for signing requests, so they can be parsed later
    
  * Update settings in docker-compose
    
  * Nonce generation and accepting a serial-request assertion instead of a serial assertion.
    
  * Refactor error handling for sign and nonce services
    
  * Fix PSQL syntax error when creating the model table
    
  * KeypairManager no longer needs the authority-id.
    snapd uses request-id to refer to the nonce, so fit in with the naming approach.
    Remove references to model revision (fixes #34).
    
  * The sign method becomes serial.
    
  * Use the serial number from the header
    
  * Log duplicate serial numbers, but allow the serial to be generated
    
  * Store the revision in the signing log
    
  * Update the version number.
    
  * Snap the vault client
    
  * 	Optimizes Docker/Docker-compose setup
    
  * Filter signing log
    
  * Substituted math lib by crypto one when generating random nonces.
    
  * See more information in merge proposal
    
  * Xpired nonces mgmt
    
  * increased digits for random generated numbers up to 10 to pass the te…
    
  * See more information in merge proposal
    
  * Model name validation
    
  * Update the installation and deployment instructions
    
  * See more information in merge proposal
    
  * See more information in merge proposal
    
  * See more information in merge proposal
    
  * Csrf prevention
    
  * Improve csrf protection
    
  * See more information in merge proposal
    
  * System user service
    
  * Composite system-user assertion
    
  * Convert the admin webapp to the vanilla-framework
    
  * Require the user key to be linked to the model
    
  * Use a different key in tests for the signing and user keys
    
  * Use the system-user key for account key.
    
  * Cache account keys
    
  * Display the account and account key assertions in the admin interface
    
  * See more information in merge proposal
    
  * Account key upload
    
  * Upload account key assertion to a keypair
    
  * New snap to install all the Serial Vault Services.
    
  * Default to HTTP and allow SSL via the 'enable-https' command
    
  * Update the basic snap to add the new system-user mode
    
  * Change to the name and location of the repository
    
  * Add Ubuntu SSO authentication with OpenID
    
  * See more information in merge proposal
    
  * See more information in merge proposal
    
  * Create tables for storing users information and the relationship with accounts
    
  * Provide roles for user authorisation
    
  * See more information in merge proposal
    
  * Display the user interface based on the user's permissions
    
  * Allow access to the admin interface when user authentication is disabled
    
  * Allow access to the admin interface if user authentication is disabled
    
  * Add the ability to logout from the application
    
  * Handle when the user unticks mandatory fields in the OpenID login
    
  * Filter the list models API based on the user
    
  * Add API permissions to the model endpoints
    
  * See more information in merge proposal
    
  * Add permissions and filtering to API endpoints
    
  * merged trunk
    
  * See more information in merge proposal
    
  * User management backend
    

server.go
// -*- Mode: Go; indent-tabs-mode: t -*-

/*
 * Copyright (C) 2016-2017 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package main

import (
	"log"
	"net/http"
	"os"

	"github.com/CanonicalLtd/serial-vault/datastore"
	"github.com/CanonicalLtd/serial-vault/service"
	"github.com/gorilla/csrf"
)

func main() {
	env := service.Env{}
	// Parse the command line arguments
	service.ParseArgs()
	err := service.ReadConfig(&env.Config, service.SettingsFile)
	if err != nil {
		log.Fatalf("Error parsing the config file: %v", err)
	}

	// Open the connection to the local database
	env.DB = datastore.OpenSysDatabase(env.Config.Driver, env.Config.DataSource)

	// Opening the keypair manager to create the signing database
	env.KeypairDB, err = service.GetKeyStore(env.Config)
	if err != nil {
		log.Fatalf("Error initializing the signing-key database: %v", err)
	}

	var handler http.Handler
	var address string

	switch service.ServiceMode {
	case "admin":
		// configure request forgery protection
		csrfSecure := true
		csrfSecureEnv := os.Getenv("CSRF_SECURE")
		if csrfSecureEnv == "disable" {
			log.Println("Disable secure flag")
			csrfSecure = false
		}

		CSRF := csrf.Protect(
			[]byte(env.Config.CSRFAuthKey),
			csrf.Secure(csrfSecure),
			csrf.HttpOnly(csrfSecure),
		)

		// Create the admin web service router
		handler = CSRF(service.AdminRouter(&env))
		address = ":8081"
	case "system-user":
		// configure request forgery protection
		csrfSecure := true
		csrfSecureEnv := os.Getenv("CSRF_SECURE")
		if csrfSecureEnv == "disable" {
			log.Println("Disable secure flag")
			csrfSecure = false
		}

		CSRF := csrf.Protect(
			[]byte(env.Config.CSRFAuthKey),
			csrf.Secure(csrfSecure),
			csrf.HttpOnly(csrfSecure),
			csrf.CookieName("XSRF-TOKEN"),
		)

		// Create the admin web service router
		if csrfSecure {
			handler = CSRF(service.SystemUserRouter(&env))
		} else {
			// Allow cross-origin access for local development
			handler = service.CORSMiddleware()(CSRF(service.SystemUserRouter(&env)))
		}
		address = ":8082"
	default:
		// Create the user web service router
		handler = service.SigningRouter(&env)
		address = ":8080"
	}

	log.Fatal(http.ListenAndServe(address, handler))
}
